x-pack/plugin/data-frame/qa/multi-node-tests/src/test/java/org/elasticsearch/xpack/dataframe/integration/DataFrameTransformIT.java

```diff
@@ -235,6 +235,49 @@ public void testContinuousDataFrameTransformUpdate() throws Exception {
         deleteDataFrameTransform(config.getId());
     }
 
+    public void testStopWaitForCheckpoint() throws Exception {
+        String indexName = "wait-for-checkpoint-reviews";
+        String transformId = "data-frame-transform-wait-for-checkpoint";
+        createReviewsIndex(indexName, 1000);
+
+        Map<String, SingleGroupSource> groups = new HashMap<>();
+        groups.put("by-day", createDateHistogramGroupSourceWithCalendarInterval("timestamp", DateHistogramInterval.DAY, null));
+        groups.put("by-user", TermsGroupSource.builder().setField("user_id").build());
+        groups.put("by-business", TermsGroupSource.builder().setField("business_id").build());
+
+        AggregatorFactories.Builder aggs = AggregatorFactories.builder()
+            .addAggregator(AggregationBuilders.avg("review_score").field("stars"))
+            .addAggregator(AggregationBuilders.max("timestamp").field("timestamp"));
+
+        DataFrameTransformConfig config = createTransformConfigBuilder(transformId,
+            groups,
+            aggs,
+            "reviews-by-user-business-day",
+            QueryBuilders.matchAllQuery(),
+            indexName)
+            .setSyncConfig(new TimeSyncConfig("timestamp", TimeValue.timeValueSeconds(1)))
+            .build();
+
+        assertTrue(putDataFrameTransform(config, RequestOptions.DEFAULT).isAcknowledged());
+        assertTrue(startDataFrameTransform(config.getId(), RequestOptions.DEFAULT).isAcknowledged());
+
+        // waitForCheckpoint: true should make the transform continue until we hit the first checkpoint, then it will stop
+        stopDataFrameTransform(transformId, false, null, true);
```

Comment:
@davidkyle you are 100% correct, there is a bad race condition here that could cause the transform to just stay running after the user called `_stop`. 

1. `onFinish` is called and has made it past the `if (transformTask.shouldStopAtCheckpoint) {` check
2. The flag is set to `true` by the user
3. `stop` is called and processed, but since the state is `INDEXING` nothing is done
4. `onFinish` completes sets the state to `STARTED`

The trick here is that the indexer transitions to `STARTED` and can get triggered again even off of failures. I think this also shows a bug in how we handle triggers to begin with. If we have not completed a checkpoint, I am not sure we should even check for changes against the indices more than once per checkpoint...

Let me mull this over

Reply:
I think something will have to be added to `doSaveState` to handle this race condition. 

When `doSaveState` is called after `onFinish` the indexer state is then `STARTED`. If the indexer state is `STARTED` and `shouldStopAtCheckpoint == true` that should give some indication of the desired behavior. Though, this may cause other issues as `doSaveState` is called when we hit an intermittent failure :(. 

More careful thought is necessary for this one.